File: patterns/java/src/refactoring_guru/patterns/singleton/example/thread_safe/Singleton.java
package refactoring_guru.patterns.singleton.example.thread_safe;

public final class Singleton {
    private static Singleton instance;
    public static String name;

    private Singleton(String name) {
        this.name = name;
    }

    public static synchronized Singleton getInstance(String name) {
        if (instance == null) {
            instance = new Singleton(name);
        }
        return instance;
    }
}

File: patterns/cs/src/RefactoringGuru/ThreadSafeSingleton/Example/Singleton.cs
using System;

namespace ThreadSafeSingleton.Example
{
    class Singleton
    {
        private static object _lock = new object();
        private static volatile Singleton instance;
        public String value;

        private Singleton(String value)
        {
            this.value = value;
        }

        public static Singleton GetInstance(String value)
        {
            if (instance == null)
            {
                lock (_lock)
                {
                    if (instance == null)
                    {
                        instance = new Singleton(value);
                    }
                }
            }
            return instance;
        }
    }
}
